### PR TITLE
DA-4598: Skip Disabled Projects

### PR DESCRIPTION
**Pull request by linonymous**
- Skip disabled projects in APIs

Signed-off-by: linonymous <linonymous@gmail.com>

### CODE REVIEW

File: affiliation/handlers.go [hunk 1]
```diff
@@ -976,6 +1067,9 @@ func Configure(api *operations.DevAnalyticsAffiliationAPI, service Service) {
 				"X-REQUEST-ID": requestID,
 			}).Info("DeleteEnrollmentsHandlerFunc: " + info)
 
+			if service.IsProjectSkipped(params.ProjectSlugs) {
```
Comment (lukaszgryglicki) [inline affiliation/handlers.go:1070]: `ProjectSlugs` can be a `,` separated list of project slugs.

**Review by lukaszgryglicki — CHANGES_REQUESTED** [review]
`ProjectSlugs` isn't just a single project slug - it was implemented to support multiple projects.
I know that we don't use it (probably?) but all APIs are implemented to support this.

**Comment by lukaszgryglicki** [comment]
I would go that way:
- Get all projects from `projectsSlug` - like `x/a,y/b,z/c` -> `['x/a', 'y/b', 'z/c']`.
- Check all of them one-by-one, exclude those disabled.
- Create `projectsSlug` back from array, for example if `y/b` is disabled that would be `x/a,z/c`.
- Proceed with the new `projectsSlugs`.
- If a new `projectSlugs` is empty - do your error handling as in this PR.

This needs more PRs: for `test`, `prod` and `master` branches.
cc @linonymous

File: affiliation/handlers.go (hunk 1 above)
Comment (linonymous) [reply affiliation/handlers.go]: didnt know this, added support for this.

**Comment by linonymous** [comment]
> I would go that way:
> 
> * Get all projects from `projectsSlug` - like `x/a,y/b,z/c` -> `['x/a', 'y/b', 'z/c']`.
> * Check all of them one-by-one, exclude those disabled.
> * Create `projectsSlug` back from array, for example if `y/b` is disabled that would be `x/a,z/c`.
> * Proceed with the new `projectsSlugs`.
> * If a new `projectSlugs` is empty - do your error handling as in this PR.
> 
> This needs more PRs: for `test`, `prod` and `master` branches. cc @linonymous

Added this algorithm. We check for all projects, skip the ones that are not there & rejoin the projects with commas & return back to handler. If the length of these projects is zero then we return an error as its forbidden

**Comment by linonymous** [comment]
image


image

image

**Comment by linonymous** [comment]
Comma separated list 
image

File: shdb/service.go [hunk 2]
```diff
@@ -9493,6 +9495,43 @@ func (s *service) GetListSlugMappings() (res *models.ListSlugMappings, err error
...
+		isSkipped = false
+		return
+	}
+	isSkipped = true
+	return
+}
+
+func (s *service) GetSkippedProjects() (res map[string]bool, err error) {
+	res = make(map[string]bool)
+	log.Info("IsProjectSkipped")
```
Comment (lukaszgryglicki) [inline shdb/service.go:9519]: `log.Info("IsProjectSkipped")` -> `log.Info("GetSkippedProjects")`.

File: shdb/service.go [hunk 3]
```diff
@@ -9493,6 +9495,43 @@ func (s *service) GetListSlugMappings() (res *models.ListSlugMappings, err error
...
+	defer func() {
+		log.Info(fmt.Sprintf("IsProjectSkipped(exit): res:%+v err:%v", res, err))
```
Comment (lukaszgryglicki) [inline shdb/service.go:9521]: `IsProjectSkipped` -> `GetSkippedProjects`.

File: shdb/service.go [hunk 4]
```diff
@@ -9493,6 +9495,43 @@ func (s *service) GetListSlugMappings() (res *models.ListSlugMappings, err error
...
+	}()
+	var ary []*models.SlugMapping
+	ary, err = s.FindSlugMappings([]string{"is_disabled"}, []interface{}{1}, false, nil)
+	if err != nil {
+		return
+	}
+	for _, item := range ary {
+		item.DaName = strings.ReplaceAll(item.DaName, "/", "-")
```
Comment (lukaszgryglicki) [inline shdb/service.go:9529]: Just wonder why we replace `/` with `-` here and in `SkipDisabledProjects`?

**Review by lukaszgryglicki — APPROVED** [review]
/lgtm, but please see my small comments.
You also need to create similar PRs for `test` and `prod` branches.

File: shdb/service.go (hunk 2 above)
Comment (linonymous) [reply shdb/service.go]: done

File: shdb/service.go (hunk 3 above)
Comment (linonymous) [reply shdb/service.go]: done

File: shdb/service.go (hunk 4 above)
Comment (linonymous) [reply shdb/service.go]: yes, I tested & it is not needed.

**Comment by lukaszgryglicki** [comment]
You can merge anytime, or let me know if you want me to merge this.

**Comment by linonymous** [comment]
> You can merge anytime, or let me know if you want me to merge this.

merged.

**Comment by lukaszgryglicki** [comment]
You need similar PRs for `test` and `prod` branch, or eventually, let me know that I need to port those changes to `test` & `prod` branches myself.

**Pull request merged**